analyze: add {Condition,Assert}CPUFeature

GiedriusS opened the pull request:
Taking a stab at implementing #14479.

Add {Condition,Assert}CPUFeature to `systemd-analyze` & friends. Implement it
by executing the cpuid instruction. Before that it is being checked how
many cpuid leaves are available to us. Add tables for common x86
features.
    
Tested via unit tests + checked that commands such as:
  
```bash
systemd-analyze condition 'AssertCPUFeature = rdrand'
```
    
Succeed as expected and that commands such as
    
```bash
systemd-analyze condition 'AssertCPUFeature = foobar'
```
    
Fail as expected.

poettering commented:
Instead of parsing /proc/cpuinfo I'd really prefer if we'd just call cpuid directly, and ifdef this out on non-x86/non-x86-64. i.e. let's use the actual APIs for that, and that's cpuid.

I.e. we can use `__get_cpuid()` which gcc provides in `<cpuid.h>`. See how `detect_vm_cpuid()` does it already, in our codebase in src/basic/virt.c. 

You just need to get CX and DX of CPUID page one. Wikipedia has a list of defined flags: 

https://en.wikipedia.org/wiki/CPUID#EAX=1:_Processor_Info_and_Feature_Bits

On cpuid page 7 are a few more flags, also see wikipedia text.

Mapping these flags to strings should be pretty straightforward.

poettering commented:
Hmm, the name ConditionCPUID suggests this would check the actual identity of the CPU. But this checks for feature flags, hence maybe name this ConditionCPUFeature= or so?

poettering commented on src/basic/virt.c:
```diff
@@ -653,6 +653,37 @@ int running_in_chroot(void) {
+        _cleanup_fclose_ FILE *f = NULL;
+        int r;
+
+        f = fopen("/proc/cpuinfo", "re");
```
note that non-x86 systems have a very very different format here...

GiedriusS commented:
> Instead of parsing /proc/cpuinfo I'd really prefer if we'd just call cpuid directly, and ifdef this out on non-x86/non-x86-64. i.e. let's use the actual APIs for that, and that's cpuid.
> 
> I.e. we can use `__get_cpuid()` which gcc provides in `<cpuid.h>`. See how `detect_vm_cpuid()` does it already, in our codebase in src/basic/virt.c.
> 
> You just need to get CX and DX of CPUID page one. Wikipedia has a list of defined flags:
> 
> https://en.wikipedia.org/wiki/CPUID#EAX=1:_Processor_Info_and_Feature_Bits
> 
> On cpuid page 7 are a few more flags, also see wikipedia text.
> 
> Mapping these flags to strings should be pretty straightforward.

So, just to be on the same page before I update this: do you think that this condition should be a no-op on non-x86/non-x86-64 ISAs? Makes sense to me, just wanted to align on this.

poettering commented:
> So, just to be on the same page before I update this: do you think that this condition should be a no-op on non-x86/non-x86-64 ISAs? Makes sense to me, just wanted to align on this.

For now, yes. I presume other archs have similar concepts, but supporting that doesn't need to be part of your PR, someone else who cares about those archs can add that.

I'd treat non x86 CPUs as CPUs with exactly zero feature flags.

poettering commented:
ping?

GiedriusS commented:
> ping?

Thanks for the ping! Was on vacation for a few weeks. I'm somewhere in the middle of reimplementing this with actual x86 instructions. Hoping to finish it up + rebase in a couple of days.

GiedriusS commented:
@poettering I have reworked this PR to actually use cpuid and renamed the flag to `{Condition,Assert}CPUFeature`. The CPUID flag tables probably need extra scrutiny. PTAL when you have some time.

poettering commented on src/basic/virt.c:
```diff
@@ -669,6 +669,132 @@ int running_in_chroot(void) {
+        const char *name;
+};
+
+static struct cpuid_table_entry leaf1_edx[] = {
```
add "const"

poettering replied:
(here and all the other arrays)

poettering commented on src/basic/virt.c:
```diff
@@ -669,6 +669,132 @@ int running_in_chroot(void) {
+        {  8, "constant_tsc" },
+};
+
+static bool given_flag_in_set(const char *flag, struct cpuid_table_entry *set, unsigned int set_size, uint32_t val) {
```
please use "size_t" for everything that counts memory objects, such as these array entries. It's also the correct type that ELEMENTSOF returns, hence please do not convert to "unsigned int" here. or in other words: `unsigned int set_size → size_t set_size`

poettering replied:
the second parameter should be const

poettering commented on src/basic/virt.c:
```diff
@@ -669,6 +669,132 @@ int running_in_chroot(void) {
+};
+
+static bool given_flag_in_set(const char *flag, struct cpuid_table_entry *set, unsigned int set_size, uint32_t val) {
+        for (unsigned int i = 0; i < set_size; i++) {
```
here too: use size_t

poettering commented on src/basic/virt.c:
```diff
@@ -669,6 +669,132 @@ int running_in_chroot(void) {
+
+static bool given_flag_in_set(const char *flag, struct cpuid_table_entry *set, unsigned int set_size, uint32_t val) {
+        for (unsigned int i = 0; i < set_size; i++) {
+                if (strcmp(flag, set[i].name) == 0 &&
```
`strcmp(…) == 0` → `streq(…)` please

poettering commented on src/basic/virt.c:
```diff
@@ -669,6 +669,132 @@ int running_in_chroot(void) {
+static bool given_flag_in_set(const char *flag, struct cpuid_table_entry *set, unsigned int set_size, uint32_t val) {
+        for (unsigned int i = 0; i < set_size; i++) {
+                if (strcmp(flag, set[i].name) == 0 &&
+                                (1ULL << set[i].flag_bit) & val)
```
1ULL is 64bit on most systems. But you use it to check against a 32bit value. 

consider using `UINT32_C(1)` instead of `1ULL`, which is a 32bt constant.

poettering commented on src/basic/virt.c:
```diff
@@ -669,6 +669,132 @@ int running_in_chroot(void) {
+                        return true;
+        }
+
+
```
spurious double empty line

poettering commented on src/basic/virt.c:
```diff
@@ -669,6 +669,132 @@ int running_in_chroot(void) {
+        }
+
+
+        if (__get_cpuid(0x80000001, &eax, &ebx, &ecx, &edx)) {
```
hmm, the constant `0x80000001` typically doesn't fit in "int" anymore, hence should come with a "U" suffix at least

poettering commented on src/basic/virt.c:
```diff
@@ -669,6 +669,132 @@ int running_in_chroot(void) {
+                        return true;
+        }
+
+        if (__get_cpuid(0x80000007, &eax, &ebx, &ecx, &edx))
```
as above, the literal should have a "U" suffix

poettering commented on src/shared/condition.c:
```diff
@@ -959,6 +969,7 @@ static const char* const assert_type_table[_CONDITION_TYPE_MAX] = {
         [CONDITION_CPUS] = "AssertCPUs",
         [CONDITION_MEMORY] = "AssertMemory",
         [CONDITION_ENVIRONMENT] = "AssertEnvironment",
+        [CONDITION_CPUFEATURE] = "AssertCPUFeature",
```
I think `CONDITION_CPU_FEATURE` would be more in line with the other naming, i.e. place an underscore between ech word

poettering reviewed (changes requested):
Looks really good, but needs some doc updates too

GiedriusS commented:
@poettering thank you for your careful review! I have addressed all of your comments. Also, I have amended the `systemd.unit` man page with information about the new condition. However, I am not completely familiar with all of the documentation sources so I might have missed something. Also, the CI failures seem completely unrelated to my changes so I believe this is ready for a third review. PTAL when you'll have some time :hugs:

mrc0mmand commented:
The aarch64 Fedora build is not happy:

```
../src/basic/virt.c:863:13: error: ‘given_flag_in_set’ defined but not used [-Werror=unused-function]
  863 | static bool given_flag_in_set(const char *flag, const struct cpuid_table_entry *set, size_t set_size, uint32_t val) {
      |             ^~~~~~~~~~~~~~~~~
../src/basic/virt.c:859:39: error: ‘leaf87_edx’ defined but not used [-Werror=unused-const-variable=]
  859 | static const struct cpuid_table_entry leaf87_edx[] = {
      |                                       ^~~~~~~~~~
../src/basic/virt.c:854:39: error: ‘leaf81_ecx’ defined but not used [-Werror=unused-const-variable=]
  854 | static const struct cpuid_table_entry leaf81_ecx[] = {
      |                                       ^~~~~~~~~~
../src/basic/virt.c:848:39: error: ‘leaf81_edx’ defined but not used [-Werror=unused-const-variable=]
  848 | static const struct cpuid_table_entry leaf81_edx[] = {
      |                                       ^~~~~~~~~~
../src/basic/virt.c:839:39: error: ‘leaf7_ebx’ defined but not used [-Werror=unused-const-variable=]
  839 | static const struct cpuid_table_entry leaf7_ebx[] = {
      |                                       ^~~~~~~~~
../src/basic/virt.c:820:39: error: ‘leaf1_ecx’ defined but not used [-Werror=unused-const-variable=]
  820 | static const struct cpuid_table_entry leaf1_ecx[] = {
      |                                       ^~~~~~~~~
../src/basic/virt.c:794:39: error: ‘leaf1_edx’ defined but not used [-Werror=unused-const-variable=]
  794 | static const struct cpuid_table_entry leaf1_edx[] = {
      |                                       ^~~~~~~~~
cc1: all warnings being treated as errors
```

(the ELN fails are "expected" and unrelated)

poettering commented on src/basic/virt.c:
```diff
@@ -786,6 +786,131 @@ int running_in_chroot(void) {
+static bool given_flag_in_set(const char *flag, const struct cpuid_table_entry *set, size_t set_size, uint32_t val) {
+        for (size_t i = 0; i < set_size; i++) {
+                if (streq(flag, set[i].name) &&
+                                (UINT32_C(1) << set[i].flag_bit) & val)
```
maybe swap the left-hand-side and right-hand-size of the `&&`, given that the bit check is likely quicker

poettering commented on src/basic/virt.c:
```diff
@@ -786,6 +786,131 @@ int running_in_chroot(void) {
+}
+
+#if defined(__i386__) || defined(__x86_64__)
+static int real_has_cpu_with_flag(const char *flag) {
```
hmm, you might as well change the return value to `bool` here, you only return true/false here anyway

poettering reviewed (changes requested):
looks dope!

please squash the two commits though (i.e. we try to merge "perfect" PRs, i.e. where thre are no fix-up commits for bugs in the same PR, but every commit is "perfect" on its own)

poettering commented:
only minor stuff found, if you fix these quickly we can add this to upcoming v248

GiedriusS commented:
@poettering I have updated the PR according to your last comments. PTAL.

poettering commented:
lgtm!

kiplingw commented:
@GiedriusS, thank you for your work. Is there any possibility you could add `avx512` and `fma4`?

GiedriusS commented:
@kiplingw could you make an issue for that? Should be straightforward to implement. I don't have as much time nowadays but maybe someone can pick it up

kiplingw commented:
No problem @GiedriusS. I just created issue #20398.
